Print out IngressRoute Status from DAG

stevesloka opened the pull request:
This updates the design for the IngressRoute `status` section which provides feedback as to how the IngressRoute object is functioning (e.g. Valid / Invalid / Orphaned).

// Fixes #449 

Signed-off-by: Steve Sloka <steves@heptio.com>

stevesloka commented:
Includes samples for applying `Valid`, `Invalid`, and `Orphaned` status, but not implemented. Those will come in future PR's.

davecheney commented on apis/contour/v1beta1/ingressroute.go:
```diff
@@ -77,6 +77,14 @@ type Delegate struct {
+type Status struct {
+	CurrentStatus   string      `json:"currentStatus"`
+	Description     string      `json:"description"`
+	LastProcessTime metav1.Time `json:"lastProcessTime"`
```
I had a spelunk in the other k8s objects looking for a theme for this field name and couldn't find a pattern. Indeed I couldn't find anything that included a timestamp in status. Even pod doesn't include the timestamp when it last updated the field, it includes a timestamp when the condition change, see https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#containerstate-v1-core

I'm wondering if this is a hint that we may be making a mistake. Including a timestamp feels like a safety blanket that if the last processed time is too old, maybe we distrust its value.

I'd like to suggest dropping last processed time for the beta.

davecheney commented on apis/contour/v1beta1/ingressroute.go:
```diff
@@ -77,6 +77,14 @@ type Delegate struct {
+	CurrentStatus   string      `json:"currentStatus"`
+	Description     string      `json:"description"`
+	LastProcessTime metav1.Time `json:"lastProcessTime"`
+	Paths           []string    `json:"paths"`
```
I thought that we weren't going to include this at the moment because the decision for beta.1 was status would be all or nothing -- everything in the document worked, or nothing in the document worked; and we'll capture that weakly in the `description:` key

stevesloka commented:
@davecheney I removed the two fields per review comments.

stevesloka commented:
fyi, I just pushed a small change where there was some unneeded code leftover and I removed.

alexbrand commented on internal/k8s/status.go:
```diff
@@ -0,0 +1,83 @@
+	if err != nil {
+		return err
+	}
+	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(existingBytes, updatedBytes, ingressroutev1.IngressRoute{})
```
I think CRDs do not support strategic merge patches:
https://github.com/kubernetes/kubernetes/issues/50037

stevesloka replied:
https://sourcegraph.com/github.com/heptio/ark@c8989231eb468f1ba9bdc531668fca8e2580c0ba/-/blob/pkg/controller/backup_deletion_controller.go#L360

stevesloka commented:
@alexbrand I rebased and updated this to use the json patch for updates. Also, copied your rbac permissions and cleaned up the design doc to match the current implementation.

stevesloka commented:
@davecheney Can you clarify your comments a bit in regards to having the "IngressRoute object being responsible for updating its own representation in the API server". 

Is this just in regards to where we do the API update or that we're storing Status in the object? 

I think once #501 lands we could have this branch process the Status list that's returned and update the K8S API accordingly in dag/k8s.go

stevesloka commented:
@alexbrand @davecheney This is updated to process through the resulting status of the Dag recompute(). I refactored the IngressRoute object to allow for storage of the dag version as well. 

The next PR I'm going to look at picking up the Lister and integrating status into the k8s objects.

davecheney commented:
Manually merged in 3d87eb6f5d60e2976975f6f8cc75f1277b166f44

Fixes #449
